File: algorithms/131. Palindrome Partitioning/solution.cc
class Solution {
public:
    vector<vector<string>> partition(string s) {
        vector<vector<string>> result;
        vector<string> path;
        dfs(0, s, path, result);
        return result;
    }

    void dfs(int p, string & s, vector<string> & path, vector<vector<string>> & result) {
        if ( s.size() == p ) {
            result.push_back(path);
            return;
        }
        for ( int i = p; i < s.size(); ++i ) {
            int l = p, r = i;
            while ( l < r && s[l] == s[r] ) {
                ++l;
                --r;
            }
            if ( l >= r ) {
                path.push_back(s.substr(p, i - p + 1));
                dfs(i + 1, s, path, result);
                path.pop_back();
            }
        }
    }
};

File: algorithms/131. Palindrome Partitioning/solution.py
class Solution(object):
    def partition(self, s):
        """
        :type s: str
        :rtype: List[List[str]]
        """
        if not s:
            return []
        res = [[s]] if s == s[::-1] else []
        for i in xrange(1, len(s)):
            ss = s[:i]
            if ss != ss[::-1]:
                continue
            for p in self.partition(s[i:]):
                res.append([ss] + p)
        return res
